File: apps/web/src/app/api/boards/_lib/localBoardSnapshotStore.ts
import { createHash, randomUUID } from 'node:crypto'
import type { Dirent } from 'node:fs'
import { mkdir, readdir, readFile, rm, writeFile } from 'node:fs/promises'
import path from 'node:path'
import { auditBoardDocument } from '@/features/boards/boardDocumentGuard'
import {
  getBoardDocumentMetrics,
  normalizeBoardThumbnailUrl,
  type BoardSnapshotCreateInput,
  type BoardSnapshotRecord,
} from '@/features/boards/boardTypes'
import type { ApiRequestContext } from '../../_lib/apiRequestContext'

const storageRoot = process.env.TANGENT_BOARD_STORAGE_DIR ?? path.join(process.cwd(), '.tangent-boards')
const snapshotLimit = Number(process.env.TANGENT_FREE_BOARD_SNAPSHOT_LIMIT ?? 100)

export async function createLocalBoardSnapshot(input: BoardSnapshotCreateInput, context: ApiRequestContext) {
  const boardId = sanitizeBoardId(input.boardId)
  if (!boardId) throw new Error('Invalid board id.')
  const audit = auditBoardDocument(input.document)
  if (!audit.ok) throw new Error(audit.issues.find((issue) => issue.blocking)?.message ?? 'Board document is blocked.')

  const createdAt = new Date().toISOString()
  const metrics = getBoardDocumentMetrics(input.document)
  const snapshot: BoardSnapshotRecord = {
    assetCount: metrics.assetCount,
    boardId,
    byteSize: audit.byteSize,
    createdAt,
    createdBy: context.userId,
    document: input.document,
    documentHash: hashDocument(input.document),
    expiresAt: null,
    id: `snapshot_${randomUUID()}`,
    reason: input.reason,
    retentionTier: 'free',
    shapeCount: metrics.shapeCount,
    thumbnailUrl: normalizeBoardThumbnailUrl(input.thumbnailUrl),
    title: input.title?.trim() || 'Untitled snapshot',
    workspaceId: context.workspaceId,
  }

  await mkdir(getSnapshotRoot(context.workspaceId, boardId), { recursive: true })
  await writeSnapshot(snapshot)
  await enforceSnapshotLimit(context, boardId)
  return summarizeSnapshot(snapshot)
}

export async function listLocalBoardSnapshots(boardId: string, context: ApiRequestContext) {
  const safeBoardId = sanitizeBoardId(boardId)
  if (!safeBoardId) throw new Error('Invalid board id.')
  let entries: Dirent[]
  try {
    entries = await readdir(getSnapshotRoot(context.workspaceId, safeBoardId), { withFileTypes: true })
  } catch (error) {
    if (isNodeError(error) && error.code === 'ENOENT') return []
    throw error
  }

  const snapshots = []
  for (const entry of entries) {
    if (!entry.isFile() || !entry.name.endsWith('.json')) continue
    try {
      const snapshot = await readSnapshot(context.workspaceId, safeBoardId, entry.name.replace(/\.json$/, ''))
      snapshots.push(summarizeSnapshot(snapshot))
    } catch {
      continue
    }
  }
  return snapshots.sort((left, right) => Date.parse(right.createdAt) - Date.parse(left.createdAt))
}

export async function loadLocalBoardSnapshot(boardId: string, snapshotId: string, context: ApiRequestContext) {
  const safeBoardId = sanitizeBoardId(boardId)
  const safeSnapshotId = sanitizeSnapshotId(snapshotId)
  if (!safeBoardId || !safeSnapshotId) throw new Error('Invalid snapshot id.')
  const snapshot = await readSnapshot(context.workspaceId, safeBoardId, safeSnapshotId)
  if (snapshot.workspaceId !== context.workspaceId || snapshot.boardId !== safeBoardId) {
    throw new Error('Board history entry not found in workspace.')
  }
  return snapshot
}

export async function clearLocalBoardSnapshots(boardId: string, context: ApiRequestContext) {
  const safeBoardId = sanitizeBoardId(boardId)
  if (!safeBoardId) throw new Error('Invalid board id.')
  const snapshots = await listLocalBoardSnapshots(safeBoardId, context)
  await rm(getSnapshotRoot(context.workspaceId, safeBoardId), { force: true, recursive: true })
  return snapshots.length
}

function summarizeSnapshot(snapshot: BoardSnapshotRecord) {
  const { document, ...summary } = snapshot
  return {
    ...summary,
    title: getSnapshotDisplayTitle(document, summary.title),
  }
}

function getSnapshotDisplayTitle(document: unknown, fallbackTitle: string) {
  const fallback = fallbackTitle.trim()
  return getKonvaActivePageTitle(document) ?? (fallback || 'Untitled snapshot')
}

function getKonvaActivePageTitle(document: unknown) {
  if (!document || typeof document !== 'object') return null
  const envelope = document as Record<string, unknown>
  if (envelope.renderer !== 'konva' || envelope.version !== 2) return null
  const activePageId = typeof envelope.activePageId === 'string' ? envelope.activePageId : 'page-1'
  const pages = Array.isArray(envelope.pages) ? envelope.pages : []
  const activePage = pages.find((page) => (
    Boolean(page)
    && typeof page === 'object'
    && (page as Record<string, unknown>).id === activePageId
  ))
  const pageTitle = activePage && typeof activePage === 'object'
    ? (activePage as Record<string, unknown>).title
    : null
  if (typeof pageTitle === 'string' && pageTitle.trim()) return pageTitle.trim()
  const canvasDocument = envelope.canvasDocument
  if (!canvasDocument || typeof canvasDocument !== 'object') return null
  const metadata = (canvasDocument as Record<string, unknown>).metadata
  if (!metadata || typeof metadata !== 'object') return null
  const documentTitle = (metadata as Record<string, unknown>).name
  return typeof documentTitle === 'string' && documentTitle.trim() ? documentTitle.trim() : null
}

async function enforceSnapshotLimit(context: ApiRequestContext, boardId: string) {
  const limit = Number.isFinite(snapshotLimit) && snapshotLimit > 0 ? snapshotLimit : 100
  const snapshots = await listLocalBoardSnapshots(boardId, context)
  const autosaves = snapshots.filter((snapshot) => getSnapshotRetentionKind(snapshot.reason) === 'autosave')
  const userSaves = snapshots.filter((snapshot) => getSnapshotRetentionKind(snapshot.reason) === 'user')
  const expired = [...autosaves.slice(limit), ...userSaves.slice(limit)]
  await Promise.all(expired.map((snapshot) => rm(getSnapshotPath(context.workspaceId, boardId, snapshot.id), { force: true })))
}

function getSnapshotRetentionKind(reason: BoardSnapshotRecord['reason']) {
  return reason === 'autosave' || reason === 'auto_interval' ? 'autosave' : 'user'
}

async function readSnapshot(workspaceId: string, boardId: string, snapshotId: string) {
  const raw = await readFile(getSnapshotPath(workspaceId, boardId, snapshotId), 'utf8')
  return JSON.parse(raw) as BoardSnapshotRecord
}

async function writeSnapshot(snapshot: BoardSnapshotRecord) {
  await writeFile(getSnapshotPath(snapshot.workspaceId, snapshot.boardId, snapshot.id), `${JSON.stringify(snapshot, null, 2)}\n`)
}

function getSnapshotRoot(workspaceId: string, boardId: string) {
  return path.join(storageRoot, 'snapshots', workspaceId, boardId)
}

function getSnapshotPath(workspaceId: string, boardId: string, snapshotId: string) {
  return path.join(getSnapshotRoot(workspaceId, boardId), `${snapshotId}.json`)
}

function hashDocument(document: unknown) {
  return createHash('sha256').update(JSON.stringify(document)).digest('hex')
}

function sanitizeBoardId(value: string | undefined) {
  if (!value) return null
  return /^[a-zA-Z0-9._-]+$/.test(value) && !value.includes('..') ? value : null
}

function sanitizeSnapshotId(value: string | undefined) {
  if (!value) return null
  return /^[a-zA-Z0-9._-]+$/.test(value) && !value.includes('..') ? value : null
}

function isNodeError(error: unknown): error is NodeJS.ErrnoException {
  return error instanceof Error && 'code' in error
}

File: services/api/tangent_api/storage/local_board_snapshot_store.py
import hashlib
import json
import os
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional
from uuid import uuid4

from fastapi import HTTPException

from tangent_api.board_guard import audit_board_document
from tangent_api.request_context import ApiRequestContext
from tangent_api.schemas import (
    BoardSnapshotCreateRequest,
    BoardSnapshotRecord,
    BoardSnapshotSummary,
    get_board_document_metrics,
)
from tangent_api.board_metadata import get_board_snapshot_display_title, normalize_board_thumbnail_url

BOARD_ID_PATTERN = re.compile(r"^[a-zA-Z0-9._-]+$")


def create_board_snapshot(
    board_id: str,
    input_data: BoardSnapshotCreateRequest,
    context: ApiRequestContext,
) -> BoardSnapshotSummary:
    safe_board_id = _sanitize_id(board_id)
    if not safe_board_id:
        raise HTTPException(status_code=400, detail="Invalid board id.")
    audit = audit_board_document(input_data.document)
    if not audit.ok:
        issue = next((item for item in audit.issues if item.blocking), None)
        raise HTTPException(status_code=422, detail=issue.message if issue else "Board document is blocked.")

    metrics = get_board_document_metrics(input_data.document)
    snapshot = BoardSnapshotRecord(
        assetCount=metrics["asset_count"],
        boardId=safe_board_id,
        byteSize=audit.byte_size,
        createdAt=datetime.now(timezone.utc).isoformat(),
        createdBy=context.user_id,
        document=input_data.document,
        documentHash=_hash_document(input_data.document),
        expiresAt=None,
        id=f"snapshot_{uuid4()}",
        reason=_normalize_reason(input_data.reason),
        retentionTier="free",
        shapeCount=metrics["shape_count"],
        thumbnailUrl=normalize_board_thumbnail_url(input_data.thumbnail_url),
        title=get_board_snapshot_display_title(input_data.document, input_data.title),
        workspaceId=context.workspace_id,
    )
    _write_snapshot(snapshot)
    _enforce_snapshot_limit(context, safe_board_id)
    return _summarize_snapshot(snapshot)


def list_board_snapshots(board_id: str, context: ApiRequestContext) -> list[BoardSnapshotSummary]:
    safe_board_id = _sanitize_id(board_id)
    if not safe_board_id:
        raise HTTPException(status_code=400, detail="Invalid board id.")
    root = _snapshot_root(context.workspace_id, safe_board_id)
    if not root.exists():
        return []
    snapshots: list[BoardSnapshotSummary] = []
    for path in root.glob("*.json"):
        try:
            snapshot = BoardSnapshotRecord.model_validate(json.loads(path.read_text(encoding="utf-8")))
        except Exception:
            continue
        snapshots.append(_summarize_snapshot(snapshot))
    return sorted(snapshots, key=lambda item: item.created_at, reverse=True)


def load_board_snapshot(
    board_id: str,
    snapshot_id: str,
    context: ApiRequestContext,
) -> BoardSnapshotRecord:
    safe_board_id = _sanitize_id(board_id)
    safe_snapshot_id = _sanitize_id(snapshot_id)
    if not safe_board_id or not safe_snapshot_id:
        raise HTTPException(status_code=400, detail="Invalid snapshot id.")
    path = _snapshot_path(context.workspace_id, safe_board_id, safe_snapshot_id)
    if not path.exists():
        raise HTTPException(status_code=404, detail="Board history entry not found.")
    snapshot = BoardSnapshotRecord.model_validate(json.loads(path.read_text(encoding="utf-8")))
    if snapshot.workspace_id != context.workspace_id or snapshot.board_id != safe_board_id:
        raise HTTPException(status_code=404, detail="Board history entry not found in workspace.")
    return snapshot


def clear_board_snapshots(board_id: str, context: ApiRequestContext) -> int:
    safe_board_id = _sanitize_id(board_id)
    if not safe_board_id:
        raise HTTPException(status_code=400, detail="Invalid board id.")
    snapshots = list_board_snapshots(safe_board_id, context)
    root = _snapshot_root(context.workspace_id, safe_board_id)
    if root.exists():
        for path in root.glob("*.json"):
            path.unlink(missing_ok=True)
    return len(snapshots)


def _write_snapshot(snapshot: BoardSnapshotRecord) -> None:
    path = _snapshot_path(snapshot.workspace_id, snapshot.board_id, snapshot.id)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(snapshot.model_dump(by_alias=True), ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def _enforce_snapshot_limit(context: ApiRequestContext, board_id: str) -> None:
    limit = _snapshot_limit()
    snapshots = list_board_snapshots(board_id, context)
    autosaves = [snapshot for snapshot in snapshots if _snapshot_retention_kind(snapshot.reason) == "autosave"]
    user_saves = [snapshot for snapshot in snapshots if _snapshot_retention_kind(snapshot.reason) == "user"]
    for snapshot in [*autosaves[limit:], *user_saves[limit:]]:
        _snapshot_path(context.workspace_id, board_id, snapshot.id).unlink(missing_ok=True)


def _summarize_snapshot(snapshot: BoardSnapshotRecord) -> BoardSnapshotSummary:
    payload = snapshot.model_dump(by_alias=True, exclude={"document"})
    payload["title"] = get_board_snapshot_display_title(snapshot.document, snapshot.title)
    return BoardSnapshotSummary(**payload)


def _snapshot_root(workspace_id: str, board_id: str) -> Path:
    return _storage_root() / "snapshots" / workspace_id / board_id


def _snapshot_path(workspace_id: str, board_id: str, snapshot_id: str) -> Path:
    return _snapshot_root(workspace_id, board_id) / f"{snapshot_id}.json"


def _storage_root() -> Path:
    return Path(os.getenv("TANGENT_BOARD_STORAGE_DIR", ".tangent-boards"))


def _hash_document(document: object) -> str:
    return hashlib.sha256(json.dumps(document, sort_keys=True, separators=(",", ":")).encode("utf-8")).hexdigest()


def _normalize_reason(value: str) -> str:
    reasons = {"autosave", "auto_interval", "keyboard", "manual", "manual_save", "pre_restore"}
    return value if value in reasons else "manual"


def _snapshot_retention_kind(reason: str) -> str:
    return "autosave" if reason in {"autosave", "auto_interval"} else "user"


def _snapshot_limit() -> int:
    try:
        value = int(os.getenv("TANGENT_FREE_BOARD_SNAPSHOT_LIMIT", "100"))
    except ValueError:
        value = 100
    return value if value > 0 else 100


def _sanitize_id(value: Optional[str]) -> Optional[str]:
    if not value:
        return None
    if BOARD_ID_PATTERN.match(value) and ".." not in value:
        return value
    raise HTTPException(status_code=400, detail="Invalid id.")
